File: packages/kotlin/src/main/kotlin/com/wave/unlock/BleTransport.kt
package com.wave.unlock

import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.delay

/** Events a BLE transport surfaces to the engine. */
sealed class BleEvent {
    data class ReaderFound(val rssi: Int) : BleEvent()
    data class Verdict(val granted: Boolean, val message: String) : BleEvent()
    object Delivered : BleEvent()
    data class Unavailable(val error: WaveError) : BleEvent()
}

/** Abstraction over the BLE stack. AndroidBleTransport is the real one; tests use MockTransport. */
interface BleTransport {
    fun events(): Flow<BleEvent>
    suspend fun write(payload: ByteArray)
    fun stop()
}

/** Scripted transport for tests and mock previews. */
class MockTransport(
    private val scripted: List<BleEvent>,
    private val interEventDelayMs: Long = 10,
) : BleTransport {
    var writtenPayload: ByteArray? = null
        private set
    @Volatile var stopped = false
        private set

    override fun events(): Flow<BleEvent> = flow {
        for (event in scripted) {
            if (stopped) break
            emit(event)
            delay(interEventDelayMs)
        }
    }

    override suspend fun write(payload: ByteArray) { writtenPayload = payload }
    override fun stop() { stopped = true }
}

File: packages/swift/Sources/WaveUnlock/BLETransport.swift
import Foundation

/// Events a BLE transport surfaces to the engine.
public enum BLEEvent: Equatable, Sendable {
    /// A reader was seen at this signal strength (may fire repeatedly).
    case readerFound(rssi: Int)
    /// Direct-BLE verdict pushed by the CBSM (status "4"/"5"), with the human message.
    case verdict(granted: Bool, message: String)
    /// Keyboard-wedge delivery ack (status "8") — credential typed out, no verdict follows.
    case delivered
    /// The transport can't proceed (BLE off, unauthorized, etc.).
    case unavailable(WaveError)
}

/// Abstraction over the BLE stack so the engine can be tested without hardware.
/// The real implementation is `CoreBluetoothTransport`; tests use `MockTransport`.
public protocol BLETransport: AnyObject, Sendable {
    /// Begin scanning; the returned stream emits `readerFound` samples and, after a
    /// successful write, any direct-BLE `verdict`/`delivered`.
    func events() -> AsyncStream<BLEEvent>
    /// Connect to the strongest candidate and write the payload (write-without-response).
    /// Must NOT wait for a notify (that deadlocks the SKBluTag).
    func write(_ payload: Data) async throws
    /// Drop the connection and stop scanning.
    func stop()
}
